syntax_and_semantics/literals/array.md

```diff
@@ -1,67 +1,82 @@
+
+Arrays are typically created with an array literal denoted by square brackets (`[]`) and individual elements separated by a comma (`,`).
 
 ```crystal
-[1, 2, 3]         # Array(Int32)
-[1, "hello", 'x'] # Array(Int32 | String | Char)
+[1, 2, 3]
 ```
 
-An Array can have mixed types, meaning `T` will be a union of types, but these are determined when the array is created, either by specifying T or by using an array literal. In the latter case, T will be set to the union of the array literal elements.
+## Generic Type Argument
 
-When creating an empty array you must always specify T:
+The array's generic type argument `T` is inferred from the types of the elements inside the literal. When all elements of the array have the same type, `T` equals to that. Otherwise it will be a union of all element types.
 
 ```crystal
-[] of Int32 # same as Array(Int32).new
-[]          # syntax error
+[1, 2, 3]          # => Array(Int32)
+[1, "hello", 'x']  # => Array(Int32 | String | Char)
 ```
 
-## Array of String
+An explicit type can be specified by immediately following the closing bracket with `of` and a type, each separated by whitespace. This overwrites the inferred type and can be used for example to create an array that holds only some types initially but can accept other types later.
+
+```crystal
+array_of_numbers = [1, 2, 3] of Number  # => Array(Number)
+array_of_numbers + [0.5]                # => [1, 2, 3, 0.5]
+
+array_of_int_or_string = [1, 3, 4] of Int32 | String  # => Array(Int32 | String)
+array_of_int_or_string + ["foo"]                      # => [1, 2, 3, "foo"]
+```
 
-Arrays of strings can be created with a special syntax:
+Empty array literals always need a type specification::
 
 ```crystal
-%w(one two three) # ["one", "two", "three"]
+[] of Int32  # => Array(Int32).new
 ```
 
-## Array of Symbol
+## Percent Array Literals
 
-Arrays of symbols can be created with a special syntax:
+[Arrays of strings](./string.html#Percent String Array Literal) and [arrays of symbols](./symbol.html#Percent Symbol Array Literal) can be created with percent array literals:
 
 ```crystal
-%i(one two three) # [:one, :two, :three]
+%w(one two three)  # => ["one", "two", "three"]
+%i(one two three)  # => [:one, :two, :three]
 ```
 
-## Array-like types
+## Array-like Type Literal
+
+Crystal supports an additional literal for arrays and array-like types. It consists of the name of the type followed by a list of elements enclosed in curly braces (`{}`) and individual elements separated by a comma (`,`).
+
+```crystal
+Array{1, 2, 3}
+```
 
-You can use a special array literal syntax with other types too, as long as they define an argless `new` method and a `<<` method:
+This literal can be used with any type as long as it has an argless constructor and responds to `<<`.
 
 ```crystal
-MyType{1, 2, 3}
+IO::Memory{1, 2, 3}
```

Comment:
Not sure this is a good example... Not sure there's a better one though.

Reply:
Agreed. I think it is better to use a type from standard lib instead of an unspecified custom type.
There aren't much options, though, espacially non-generic ones.